AWSIoT/Internal/AWSIoTMQTTClient.m

```diff
@@ -706,13 +739,19 @@ - (void)reconnectToSession {
 
 - (void)notifyConnectionStatus {
     //Set the connection status on the callback.
-    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
-        if (self.connectStatusCallback != nil) {
-            self.connectStatusCallback(self.mqttStatus);
+    AWSIoTMQTTStatus mqttStatus = self.mqttStatus;
+    __weak AWSIoTMQTTClient *weakSelf = self;
+    __weak StatusCallback connectStatusCallback = self.connectStatusCallback;
+    __weak id<AWSIoTMQTTClientDelegate> clientDelegate = self.clientDelegate;
+    dispatch_barrier_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
```

Comment:
can we have a unit test for the callback changes, maybe something like the issue where we call `connect()`, `disconnect()` and `connect()` and verify the old callback is not fired?

Reply:
Done!